### ISSUE TITLE
Variable number of arguments

### ISSUE THREAD

**Issue by monolithed** [KyleAMathews/deepmerge#33]
It would be nice if deepmerge will support a viariable number of arguments

**Comment by TehShrike** [comment]
That would make sense to me, too.  It might be a bit trickier to add though, since so many people seem to want the ability to configure array merging via a third argument (#14).

It's probably still possible, though.  I wouldn't be against a pull request.

**Comment by TehShrike** [comment]
If anyone wants to suggest a backwards-compatible API and/or open a pull request, fire away.

**Comment by declandewet** [comment]
While this option may be convenient, it's not too difficult to just write a wrapper around `deepmerge` on a case-by-case basis:

```js
const deepMerge = require("deepmerge")

const deepMergeAll = (...xs) => {
  const initial = xs.shift()
  return xs
    .filter((x) => typeof x !== 'undefined' && x !== null)
    .reduce((x, y) => deepMerge(x, y), initial || {})
}

deepMergeAll({ a: 1 }, { b: 2 }, null, { c: 3 })
// { a: 1, b: 2, c: 3 }
```

I don't think there are very many projects that use more than a maximum of 5 different array merge strategies (that just sounds super rare & unrealistic, imo) so as far as boilerplate goes, this isn't too bad.

**Comment by Bloomca** [comment]
Well, it is possible just to expose another function, as @declandewet suggested.

**Comment by TehShrike** [comment]
How do y'all feel about an API like `const merged = deepmerge.options({ clone: true })(a, b, c, d, e)`?

```js
const mergeAll = deepmerge.options({ arrayMerge: someFunction })

const mergedA = mergeAll(a, b, c)
const mergedB = mergeAll(b, c, a)
```

**Comment by monolithed** [comment]
My suggestion:

```
merge([x, y, z], { options: { ... } });
```

**Comment by TehShrike** [comment]
I like that, though I don't think we could distinguish between that and somebody trying to merge a regular object into a normal array.  We'd have to add a new method, `merge.all([x, y, z], { /* options */ })` or something.

**Comment by declandewet** [comment]
You can't merge a Boolean, so you could inspect if the first argument is boolean and use that as a hint that the 2nd param is an options object and return a function that merges all arguments

```js
const deepMergeAll =  deepmerge(true, [options])

const merged = deepMergeAll(a, b, c, d)
```

**Comment by TehShrike** [comment]
Adding a new function like `merge.all` or `merge.options` wouldn't be a breaking change, just a feature version bump.

Relying on special values in ordered arguments makes me nervous.

**Comment by Bloomca** [comment]
I feel the same as @TehShrike – there is no particular need to break things, and passing `Boolean` looks very esoteric. I feel that new function is the most "soft" solution here.

Though I am not sure about instantiating new function with different merge strategies – it looks kind of strange too (at least for me). I can't really suggest better solution – only the first argument, and then collect other values through rest spread, but then we make first argument mandatory.

**Comment by TehShrike** [comment]
Vote for your preference via reactions (assume that the names are changeable, so vote just based on api/functionality):

Option ❤️ 
```js
const merge = require('deepmerge')

const mergeAll = merge.options({ clone: true })

const mergedObject = mergeAll(a, b, c)
```

Option 🎉 
```js
const merge = require('deepmerge')

const mergedObject = merge.all([a, b, c], { clone: true })
```

**Comment by TehShrike** [comment]
Sounds good to me!  We'll go with `merge.all([a, b, c], { clone: true })`.

Unless anyone thinks of an objection, I'll say the specs are something like:
- Options object should be optional
- An error should be thrown if the first argument is not an array of at least two elements

Pull requests with tests for these checks, or tests for the rest of the functionality, would be much appreciated.  I think they should go into a new file `test/merge-all.js`.

**Comment by Bloomca** [comment]
@TehShrike I'd like to do this! I can do this on weekend.

**Comment by macdja38** [comment]
This sounds Like a very good way to go about this!

**Comment by TehShrike** [comment]
This is implemented by #50 and published as 1.3.0!
